### ISSUE TITLE
terraform lock never releases

### ISSUE THREAD

**Issue by sgendler-stem** [hashicorp/terraform#16652]
I have a template with a lock that seems to be perma-locked.  For several days, it has just blocked on "Acquiring state lock. This may take a few moments..."

However, I have NEVER run anything without -lock-timeout=20m in the command.  However, a command did get interrupted by network loss on the day when the lock got locked.  I'm guessing acquiring the lock and setting the timeout are somehow not part of the same atomic operation?  I'm using the following in my top-level terraform.tfvars file:

```
  remote_state {
    backend = "s3"

    config {
      bucket = "<redacted>"
      key    = "${path_relative_to_include()}/terraform.tfstate"
      region = "us-west-2"
      encrypt = true
      dynamodb_table = "terraform-lock-table"
    }

  terraform {
    extra_arguments "common_vars" {
      commands = ["${get_terraform_commands_that_need_vars()}"]

      optional_var_files = [
        "${get_tfvars_dir()}/${find_in_parent_folders("account.tfvars", "skip-account-if-does-not-exist")}",
        "${get_tfvars_dir()}/${find_in_parent_folders("region.tfvars", "skip-region-if-does-not-exist")}",
        "${get_tfvars_dir()}/${find_in_parent_folders("env.tfvars", "skip-env-if-does-not-exist")}",
        "${get_tfvars_dir()}/terraform.tfvars"
      ]
    }

    extra_arguments "retry_lock" {
      commands = ["${get_terraform_commands_that_need_locking()}"]
      arguments = ["-lock-timeout=20m"]
    }
  }
}
```

What is particularly obnoxious is that I can't even exit the process normally.  It captures the ctl-C and just swallows it while it blocks on the lock:

```
Samuels-MBP:bastion sgendler$ terragrunt plan --terragrunt-source ../../../../../stem-infra//bastion/  
[terragrunt] [/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion] 2017/11/14 14:37:51 Running command: terraform --version
[terragrunt] 2017/11/14 14:37:51 Reading Terragrunt config file at /Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion/terraform.tfvars
[terragrunt] 2017/11/14 14:37:51 Cleaning up existing *.tf files in /var/folders/xr/t6gsrby97350k0r85qr7blzh0000gn/T/terragrunt/Fxxb1YsbNFq0VqlPDqwElrdbe-4/N7YR7JXFv_AHrQ_vUpC9GGTlLbM
[terragrunt] 2017/11/14 14:37:51 Downloading Terraform configurations from file:///Users/sgendler/src/stem/stem-infra into /var/folders/xr/t6gsrby97350k0r85qr7blzh0000gn/T/terragrunt/Fxxb1YsbNFq0VqlPDqwElrdbe-4/N7YR7JXFv_AHrQ_vUpC9GGTlLbM using terraform init
[terragrunt] [/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion] 2017/11/14 14:37:51 Backend s3 has not changed.
[terragrunt] [/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion] 2017/11/14 14:37:52 Running command: terraform init -backend-config=bucket=stem-terraform-state-bucket -backend-config=key=aws2/us-east-1/_global/bastion/terraform.tfstate -backend-config=region=us-west-2 -backend-config=encrypt=true -backend-config=dynamodb_table=terraform-lock-table -lock-timeout=20m -from-module=file:///Users/sgendler/src/stem/stem-infra /var/folders/xr/t6gsrby97350k0r85qr7blzh0000gn/T/terragrunt/Fxxb1YsbNFq0VqlPDqwElrdbe-4/N7YR7JXFv_AHrQ_vUpC9GGTlLbM
Copying configuration from "file:///Users/sgendler/src/stem/stem-infra"...
Terraform initialized in an empty directory!

The directory has no Terraform configuration files. You may begin working
with Terraform immediately by creating Terraform configuration files.
[terragrunt] 2017/11/14 14:37:52 Copying files from /Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion into /var/folders/xr/t6gsrby97350k0r85qr7blzh0000gn/T/terragrunt/Fxxb1YsbNFq0VqlPDqwElrdbe-4/N7YR7JXFv_AHrQ_vUpC9GGTlLbM/bastion
[terragrunt] 2017/11/14 14:37:52 Setting working directory to /var/folders/xr/t6gsrby97350k0r85qr7blzh0000gn/T/terragrunt/Fxxb1YsbNFq0VqlPDqwElrdbe-4/N7YR7JXFv_AHrQ_vUpC9GGTlLbM/bastion
[terragrunt] 2017/11/14 14:37:52 Backend s3 has not changed.
[terragrunt] 2017/11/14 14:37:52 Running command: terraform plan -var-file=/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion/../../../account.tfvars -var-file=/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion/../../region.tfvars -var-file=/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion/../env.tfvars -var-file=/Users/sgendler/src/stem/stem-envs/aws2/us-east-1/_global/bastion/terraform.tfvars -lock-timeout=20m
Acquiring state lock. This may take a few moments...
^C[terragrunt] 2017/11/14 14:41:35 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:36 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:37 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:37 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:37 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:40 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:41 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:41 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:41 Forward signal interrupt to terraform.
^C[terragrunt] 2017/11/14 14:41:41 Forward signal interrupt to terraform.
```

If I reduce the timeout to something more reasonable, it errors out after the timeout, and offers advice for running with -lock=false, but doing so doesn't actually release the lock.  It simply doesn't bother to check it.  How do I actually remove the lock so that normal life can go on?  I've seen nothing in any of the documentation that I've read which suggests that there is a way to remove a lock which won't unlock.  Should't locks auto-expire after a period, at a minimum?  That's actually what I figured the lock timeout value was for, not how long it waits for the lock to be released by another process.

**Comment by sgendler-stem** [comment]
In the end, I just deleted the key in the dynamodb table in the AWS console, but it would have been a heck of a lot easier and faster with some documentation about how to resolve this somewhere.

**Comment by jbardin** [comment]
Hi @sgendler-stem,

Sorry this was an issue for you. The command to fix this from the cli is `terraform force-unlock`, but  in the s3 case, removing the DynamoDB entry has the same effect. 

Yes, the `-lock-timeout` is the `Duration to retry a state lock`; any expiration on the lock is an implementation detail of the specific backend, for consul state locks are tied to the running process.

Interrupting the process however should have aborted the lock attempt, which I'll look into as well.
